### PR TITLE
Merge upstream

### PR DESCRIPTION
**Pull request by acomodi**
I still need to check `all_xc7` target in SymbiFlow

### CODE REVIEW

**Review by litghost — APPROVED** [review]
LGTM, if all_xc7 looks good with this PR

**Comment by acomodi** [comment]
@litghost Between current master and this PR there have been few changes in VPR (mostly related to routing).
I need to analyze what is causing this routing issue.

I have doubled checked and `counter_arty` is working with current SymbiFlow master (5f16394fda395caa93f3629034528714f1fb7f3e).

This is the error:

```	 
This is possible on a fringe node based on low Fc_out, N, and certain lengths.
## Loading routing resource graph took 11.31 seconds (max_rss 2898.4 MiB, delta_rss +35.0 MiB)
  RR Graph Nodes: 1538497
  RR Graph Edges: 4738718
Confirming router algorithm: TIMING_DRIVEN.
---- ------ ------- ---- ------- ------- ------- ----------------- --------------- -------- ---------- ---------- ---------- ---------- --------
Iter   Time    pres  BBs    Heap  Re-Rtd  Re-Rtd Overused RR Nodes      Wirelength      CPD       sTNS       sWNS       hTNS       hWNS Est Succ
      (sec)     fac Updt    push    Nets   Conns                                       (ns)       (ns)       (ns)       (ns)       (ns)     Iter
---- ------ ------- ---- ------- ------- ------- ----------------- --------------- -------- ---------- ---------- ---------- ---------- --------
Cannot route from SYN-GND.GND[0] (RR node: 113448 type: SOURCE location: (35,26) class: 0) to BLK-TL-SLICEL.CX[0] (RR node: 203426 type: SINK location: (53,29) class: 22) -- no possible path
Failed to route connection from '$false' to '$auto$alumacc.cc:474:replace_alu$6.slice[4].carry4_full' for net '$false' (#2)
Routing failed.
# Routing took 12.72 seconds (max_rss 2898.4 MiB, delta_rss +35.0 MiB)
Circuit is unroutable with a channel width factor of 500.
VPR failed to implement circuit
The entire flow of VPR took 24.26 seconds (max_rss 2898.4 MiB)
xc7/tests/counter/CMakeFiles/counter_arty_bin.dir/build.make:88: recipe for target 'xc7/tests/counter/counter_arty/artix7-xc7a50t-arty-roi-virt-xc7a50t-arty-test/top.route' failed
make[3]: *** [xc7/tests/counter/counter_arty/artix7-xc7a50t-arty-roi-virt-xc7a50t-arty-test/top.route] Error 2
make[3]: *** Deleting file 'xc7/tests/counter/counter_arty/artix7-xc7a50t-arty-roi-virt-xc7a50t-arty-test/top.route'
CMakeFiles/Makefile2:62431: recipe for target 'xc7/tests/counter/CMakeFiles/counter_arty_bin.dir/all' failed
make[2]: *** [xc7/tests/counter/CMakeFiles/counter_arty_bin.dir/all] Error 2
CMakeFiles/Makefile2:62438: recipe for target 'xc7/tests/counter/CMakeFiles/counter_arty_bin.dir/rule' failed
make[1]: *** [xc7/tests/counter/CMakeFiles/counter_arty_bin.dir/rule] Error 2
Makefile:19857: recipe for target 'counter_arty_bin' failed
make: *** [counter_arty_bin] Error 2
```

**Comment by litghost** [comment]
You can sanity check the routing failure with https://github.com/SymbiFlow/symbiflow-arch-defs/pull/680/ to see if the underlying graph is broken, or if VPR just failed to identify the route.

**Comment by litghost** [comment]
Given that https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/517 was just merged, maybe merge from there?

**Comment by acomodi** [comment]
Makes sense

**Comment by acomodi** [comment]
@litghost By using the `rr_graph_walk.py` from https://github.com/SymbiFlow/symbiflow-arch-defs/pull/680 a route could be found from the failing (I have used start and end node from the failure output).

I have narrowed down the issue which is in the recent commits regarding routing: from https://github.com/SymbiFlow/vtr-verilog-to-routing/pull/59/commits/09d365c2d287ca69f5a726b2aacb40c444871e1d to https://github.com/SymbiFlow/vtr-verilog-to-routing/pull/59/commits/616833ee243d03fd9a927296b6ff2a851a691923. These are the only changes regarding the router. I still need to analyze the issue further.

**Comment by acomodi** [comment]
I have merged the auto-format update. There still is the need to fix the routing issue.

**Comment by acomodi** [comment]
@litghost I have temporary reverted the router changes that prevented to correctly build of some of the symbiflow tests. I have filed an issue to keep track of the revert (https://github.com/SymbiFlow/vtr-verilog-to-routing/issues/61).

**Comment by acomodi** [comment]
@litghost I am running `all_xc7` on symbiflow. I´lI merge if it succeeds

**Comment by acomodi** [comment]
A new error showed up related to the virtual `rr_graph` for the arty device (during placement there is a timing and delay cost check that fails):

```
## Computing placement delta delay look-up took 1.04 seconds (max_rss 96.8 MiB, delta_rss +0.0 MiB)

There are 2 point to point connections in this circuit.


BB estimate of min-dist (placement) wire length: 130

Completed placement consistency check successfully.
Initial placement cost: 1 bb_cost: 21.6667 td_cost: 1.20218e-08 delay_cost: 1.20218e-08
Initial placement estimated Critical Path Delay (CPD): 12.1098 ns
Initial placement estimated setup Total Negative Slack (sTNS): -12.1098 ns
Initial placement estimated setup Worst Negative Slack (sWNS): -12.1098 ns

Initial placement estimated setup slack histogram:
[ -1.2e-08: -1.2e-08) 1 (100.0%) |**************************************************
[ -1.2e-08: -1.2e-08) 0 ( 0.0%) |
[ -1.2e-08: -1.2e-08) 0 ( 0.0%) |
[ -1.2e-08: -1.2e-08) 0 ( 0.0%) |
[ -1.2e-08: -1.2e-08) 0 ( 0.0%) |
[ -1.2e-08: -1.2e-08) 0 ( 0.0%) |
[ -1.2e-08: -1.2e-08) 0 ( 0.0%) |
[ -1.2e-08: -1.2e-08) 0 ( 0.0%) |
[ -1.2e-08: -1.2e-08) 0 ( 0.0%) |
[ -1.2e-08: -1.2e-08) 0 ( 0.0%) |

------- ------- ---------- ---------- ---------- ---------- ------- ---------- -------- ------- ------- ------- ------ --------- ------
      T    Cost Av BB Cost Av TD Cost Av Tot Del P to P Del     CPD       sTNS     sWNS Ac Rate Std Dev R limit    Exp Tot Moves  Alpha
------- ------- ---------- ---------- ---------- ---------- ------- ---------- -------- ------- ------- ------- ------ --------- ------
Warning 2442: Starting t: 1 of 3 configurations accepted.
  0.000  0.5067    10.0000 6.837e-09  6.837e-09  3.4185e-09   6.925      -6.92   -6.925  0.2500  0.0000 157.0000  1.000        4  0.000
Error 1: timing_cost_check: 7.15031e-09 and timing_cost: 6.83695e-09 differ in check_place.
Error 2: delay_cost_check: 7.15031e-09 and delay_cost: 6.83695e-09 differ in check_place.
Error 3: 
Type: Placement
File: /home/build/acomodi/vtr-antmicro/vpr/src/place/place.cpp
Line: 3526
Message: 
Completed placement consistency check, 2 errors found.
Aborting program.


BB estimate of min-dist (placement) wire length: 60
# Placement took 1.04 seconds (max_rss 96.8 MiB, delta_rss +0.0 MiB)
The entire flow of VPR took 6.89 seconds (max_rss 96.8 MiB)
xc7/archs/artix7/devices/CMakeFiles/file_xc7_archs_artix7_devices_rr_graph_xc7a50t-arty_test.rr_graph.virt.xml.dir/build.make:63: recipe for target 'xc7/archs/artix7/devices/rr_graph_xc7a50t-arty_test.rr_graph.virt.xml' failed
make[3]: *** [xc7/archs/artix7/devices/rr_graph_xc7a50t-arty_test.rr_graph.virt.xml] Error 1

```

**Comment by litghost** [comment]
> A new error showed up related to the virtual `rr_graph` for the arty device (during placement there is a timing and delay cost check that fails):

I've seen this before, but I'm surprised it cropped up again.  It is probably worth sending this issue upstream, but also for now lets just demote that check to a warning.

This is a good example of where the configurable throw/warning system would come in handy.

**Comment by acomodi** [comment]
Ok, probably I have incorrectly solved a merge conflict and upgraded the WARN to THROW. 
I'll demote it again, re-run the tests and, if hopefully no more errors show up, I'll land this PR

**Comment by litghost** [comment]
> Ok, probably I have incorrectly solved a merge conflict and upgraded the WARN to THROW.
> I'll demote it again, re-run the tests and, if hopefully no more errors show up, I'll land this PR

Probably: https://github.com/SymbiFlow/vtr-verilog-to-routing/commit/63eb02002723a6d57ffa28fb5e197c15d071a9fc#diff-8d28601fd7699a551f46045cd594eedc

**Comment by acomodi** [comment]
@litghost  There was an addition of a `--strict_checks` in upstream VTR that does demote some of those checks to warnings. I will open a PR in arch-defs to include that option and it can be merged after we land this PR (and conda is updated)

**Comment by acomodi** [comment]
`all_xc7` tests passed and also Travis CI. Merging

**Pull request merged**
